alonzo/impl/src/Cardano/Ledger/Alonzo/TxWitness.hs

```diff
@@ -0,0 +1,289 @@
+    _witsBoot :: Set (BootstrapWitness era),
+    _witsScriptData :: ScriptData era
+  }
+  deriving (Generic, Typeable)
+
+newtype TxWitness era = TxWitnessConstr (MemoBytes (TxWitnessRaw era))
+
+pattern TxWitness ::
+  (Era era, ToCBOR (Core.Script era)) =>
+  Set (WitVKey 'Witness era) ->
+  Set (BootstrapWitness era) ->
+  Map (ScriptHash era) (Core.Script era) ->
+  Set (Data era) ->
+  Map RdmrPtr (Data era) ->
+  TxWitness era
+pattern TxWitness
+  { witsVKey,
+    witsBoot,
+    witsScript,
+    witsData,
+    witsRdmr
+  } <-
+  TxWitnessConstr
+    ( Memo
+        ( TxWitnessRaw
+            witsVKey
+            witsBoot
+            ( ScriptDataConstr
+                ( Memo
+                    ( ScriptDataRaw
+                        witsScript
+                        witsData
+                        witsRdmr
+                      )
+                    _
+                  )
+              )
+          )
+        _
+      )
+  where
+    TxWitness witsVKey witsBoot witsScript witsDat witsRdmr =
+      TxWitnessConstr
+        . memoBytes
+        $ encodeWitnessRaw witsVKey witsBoot witsScript witsDat witsRdmr
+
+-- | Right-biased semigroup - if there are (somehow) multiple entries either for
+-- a given 'ScriptHash' or a given 'Data', this will bias to the entry on the
+-- right. Note that this should not happen in practise.
+instance
+  (Era era, ToCBOR (Core.Script era)) =>
+  Semigroup (TxWitness era)
+  where
+  TxWitness a b c d e <> TxWitness a' b' c' d' e' =
+    TxWitness
+      (a `Set.union` a')
+      (b `Set.union` b')
+      (c <> c')
+      (d `Set.union` d')
+      (e <> e')
+
+instance
+  (Era era, ToCBOR (Core.Script era)) =>
+  Monoid (TxWitness era)
+  where
+  mempty = TxWitness mempty mempty mempty mempty mempty
+
+--------------------------------------------------------------------------------
+-- Accessors
+--------------------------------------------------------------------------------
+instance
+  (Core.Script era ~ script) =>
+  HasField "witsScript" (TxWitness era) (Map (ScriptHash era) script)
+  where
+  getField (TxWitnessConstr (Memo m _)) =
+    _scriptDataScripts
+      . memotype
+      . unScriptDataConstr
+      $ _witsScriptData m
+
+instance HasField "witsData" (TxWitness era) (Set (Data era)) where
+  getField (TxWitnessConstr (Memo m _)) =
+    _scriptDataData
+      . memotype
+      . unScriptDataConstr
+      $ _witsScriptData m
+
+instance HasField "witsRdmrs" (TxWitness era) (Map RdmrPtr (Data era)) where
+  getField (TxWitnessConstr (Memo m _)) =
+    _scriptDataRdmrs
+      . memotype
+      . unScriptDataConstr
+      $ _witsScriptData m
+
+--------------------------------------------------------------------------------
+-- ScriptData
+--------------------------------------------------------------------------------
+
+data ScriptDataRaw era = ScriptDataRaw
+  { _scriptDataScripts :: Map (ScriptHash era) (Core.Script era),
+    _scriptDataData :: Set (Data era),
```

Comment:
Using the correct plural of Datum here seems slightly more confusing than using the wrong one to make it clear that this is a collection of Datum objects (Datums) heeehee

Reply:
also, should we make the scriptDataData and scriptDataScripts consistent with each other - as in, either have them both be a map from hash to the preimage, or have them both be just the preimage? And if so, which way do you prefer?

Reply:
I'm fine with either. The spec currently has them just being the preimage, but the existing Shelley code has them as the map. As yet I haven't grokked the usage patterns well enough to proffer an opinion on what would be best for `Data`. If we want them consistent, then I can happily change this to a Map.